File: SampleXamarin/SampleiOS/SetupViewController.cs
using System;
using Foundation;
using UIKit;
using CoreFoundation;

namespace HelpLightning.SDK.Sample.iOS
{
    public partial class SetupViewController : UIViewController
    {
        private static readonly string DefaultContactEmail = "small_u11@helplightning.com";

        public SetupViewController(IntPtr handle) : base(handle)
        {
        }

        public override void ViewDidLoad()
        {
            base.ViewDidLoad();
            // Perform any additional setup after loading the view, typically from a nib.
            authTokenTextField.Text = CallManager.Instance.AuthToken;
            contactEmailTextField.Text = DefaultContactEmail;
        }

        public override void DidReceiveMemoryWarning()
        {
            base.DidReceiveMemoryWarning();
            // Release any cached data, images, etc that aren't in use.
        }

        public override bool ShouldPerformSegue(string segueIdentifier, NSObject sender)
        {
            return CallManager.Instance.SessionID != null && CallManager.Instance.SessionToken != null && CallManager.Instance.UserToken != null; ;
        }

        partial void OnCreateCall(UIButton sender)
        {
            var email = contactEmailTextField.Text.Trim();
            if (email.Length == 0)
            {
                Console.WriteLine("Empty Contact Email");
                return;
            }

            indicator.Hidden = false;
            createCallButton.Enabled = false;
            getCallButton.Enabled = false;

            CallManager.Instance.ContactEmail = email;
            CallManager.Instance.SessionID = null;
            CallManager.Instance.SessionToken = null;
            CallManager.Instance.UserToken = null;

            try
            {
                var task = HLServerClient.Instance.CreateCall(CallManager.Instance.AuthToken, email);
                task.ContinueWith(t => {
                    DispatchQueue.MainQueue.DispatchAsync(() =>
                    {
                        if (t.IsCompletedSuccessfully)
                        {
                            var json = task.Result;
                            CallManager.Instance.SessionID = json["session_id"].ToString();
                            CallManager.Instance.SessionToken = json["session_token"].ToString();
                            CallManager.Instance.UserToken = json["user_token"].ToString();
                            CallManager.Instance.GssServerURL = json["url"].ToString();
                            CallManager.Instance.SessionPIN = json["sid"].ToString();

                            PerformSegue("segueOpenJoinScreen1", this);
                        }
                        else
                        {
                            Console.Error.WriteLine("Cannot Create a Call: " + t.Exception);
                        }
                        createCallButton.Enabled = true;
                        getCallButton.Enabled = true;
                        indicator.Hidden = true;
                    });
                });
            }
            catch (Exception e)
            {
                Console.Error.WriteLine(e);
            }
        }

        partial void OnGetCall(UIButton sender)
        {
            var pin = PINTextField.Text.Trim();
            if (pin.Length == 0)
            {
                Console.WriteLine("Empty PIN");
                return;
            }
            CallManager.Instance.SessionPIN = pin;
            CallManager.Instance.SessionID = null;
            CallManager.Instance.SessionToken = null;
            CallManager.Instance.UserToken = null;
            try
            {
                var task = HLServerClient.Instance.GetCall(pin, CallManager.Instance.AuthToken);
                task.ContinueWith(t =>
                {
                    DispatchQueue.MainQueue.DispatchAsync(() =>
                    {
                        if (t.IsCompletedSuccessfully)
                        {
                            var json = task.Result;
                            CallManager.Instance.SessionID = json["session_id"][0].ToString();
                            CallManager.Instance.SessionToken = json["session_token"].ToString();
                            CallManager.Instance.UserToken = json["user_token"].ToString();
                            CallManager.Instance.GssServerURL = json["url"][0].ToString();

                            PerformSegue("segueOpenJoinScreen2", this);
                        }
                        else
                        {
                            Console.Error.WriteLine("Cannot Create a Call: " + t.Exception);
                        }
                        createCallButton.Enabled = true;
                        getCallButton.Enabled = true;
                        indicator.Hidden = true;
                    });
                });
            }
            catch (Exception e)
            {
                Console.Error.WriteLine(e);
            }
        }

        partial void OnCancel(UIBarButtonItem sender)
        {
            this.PresentingViewController.DismissViewController(true, null);
        }
    }

}

File: SampleSwift/SampleSwift/SetupViewController.m
/* -*- Mode:objc; c-basic-offset: 4; indent-tabs-mode: nil; -*- */
/*
 **
 ** Copyright (c) 2020 VIPAAR, LLC all rights reserved.
 **
 ** Any commercial use of this software is subject to license
 ** agreement. Disclosure of the information contained herein to any
 ** third parties is prohibited without prior written consent. This
 ** software is subject to US export control regulations.
 **
 ** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 ** ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 ** LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 ** FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 ** AUTHORS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 ** INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 ** (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 ** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 ** INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 ** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 ** NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 ** THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 **
 */


#import "SetupViewController.h"
#import "HLServerClient.h"
#import "CallManager.h"

NSString* const kDefaultContactEmail = @"[YOUR_USER_NAME]";

@interface SetupViewController ()
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *indicator;
@property (weak, nonatomic) IBOutlet UIButton *createCallButton;
@property (weak, nonatomic) IBOutlet UIButton *getCallButton;
@property (nonatomic, weak) IBOutlet UITextField *authTokenTextField;
@property (nonatomic, weak) IBOutlet UITextField *contactEmailTextField;
@property (nonatomic, weak) IBOutlet UITextField *PINTextField;

@end

@implementation SetupViewController

- (void) viewDidLoad {
    [super viewDidLoad];
    self.contactEmailTextField.text = kDefaultContactEmail;
    self.authTokenTextField.text = CallManager.sharedInstance.authToken;
}

- (BOOL)shouldPerformSegueWithIdentifier:(NSString *)identifier sender:(id)sender {
    return CallManager.sharedInstance.sessionID != nil
    && CallManager.sharedInstance.sessionToken != nil
    && CallManager.sharedInstance.userToken != nil;
}

- (IBAction)OnCancel:(UIBarButtonItem *)sender {
    [self.presentingViewController dismissViewControllerAnimated:YES completion:nil];
}

- (IBAction)OnCreateCall:(UIButton *)sender {
    NSString* contactEmail = self.contactEmailTextField.text;
    if (contactEmail.length == 0) {
        NSLog(@"No contact email");
        return;
    }
    self.indicator.hidden = NO;
    self.createCallButton.enabled = NO;
    self.getCallButton.enabled = NO;
    
    CallManager.sharedInstance.contactEmail = contactEmail;
    
    CallManager.sharedInstance.sessionID = nil;
    CallManager.sharedInstance.sessionToken = nil;
    CallManager.sharedInstance.userToken = nil;
    
    [HLServerClient.sharedInstance createCallWithAuthToken:CallManager.sharedInstance.authToken contactEmail:contactEmail].then(^id(id result) {
        NSDictionary* json = (NSDictionary*)result;
        CallManager.sharedInstance.sessionID = json[@"session_id"];
        CallManager.sharedInstance.sessionToken = json[@"session_token"];
        CallManager.sharedInstance.userToken = json[@"user_token"];
        CallManager.sharedInstance.gssServerURL = json[@"url"];
        CallManager.sharedInstance.sessionPIN = json[@"sid"];
        
        [self performSegueWithIdentifier:@"segueOpenJoinScreen1" sender:self];
        return result;
    }).catch(^(NSError* error) {
        NSLog(@"%@", error);
    }).always(^() {
        self.indicator.hidden = YES;
        self.createCallButton.enabled = YES;
        self.getCallButton.enabled = YES;
    });
    
}

- (IBAction) OnGetCall:(UIButton *)sender {
    NSString* pin = self.PINTextField.text;
    if (pin.length == 0) {
        NSLog(@"No PIN");
        return;
    }
    self.indicator.hidden = NO;
    self.createCallButton.enabled = NO;
    self.getCallButton.enabled = NO;
    
    CallManager.sharedInstance.sessionPIN = pin;
    
    CallManager.sharedInstance.sessionID = nil;
    CallManager.sharedInstance.sessionToken = nil;
    CallManager.sharedInstance.userToken = nil;
    
    [HLServerClient.sharedInstance retreiveCallWithAuthToken:CallManager.sharedInstance.authToken
                                                         pin:pin]
    .then(^id(id result) {
        NSDictionary* json = (NSDictionary*)result;
        CallManager.sharedInstance.sessionID = json[@"session_id"];
        CallManager.sharedInstance.sessionToken = json[@"session_token"];
        CallManager.sharedInstance.userToken = json[@"user_token"];
        CallManager.sharedInstance.gssServerURL = json[@"url"];
        [self performSegueWithIdentifier:@"segueOpenJoinScreen2" sender:self];
        return result;
    }).catch(^(NSError* error) {
        NSLog(@"%@", error);
    }).always(^() {
        self.indicator.hidden = YES;
        self.createCallButton.enabled = YES;
        self.getCallButton.enabled = YES;
    });
}

@end
